### ISSUE TITLE
Unauthorized access on secondary control plane context

### ISSUE THREAD

**Issue by MSpencer87** [rancher/rancher#36518]
**Environmental Info:**

Provider: Amazon EC2 
Instance: t3.xlarge
AMI: ubuntu-20-docker-20.07
Tools: RKE v1.3.3
Release: Rancher v2.6.3
Kubernetes Version: v1.21.7 (Default)

Cluster Configuration:

2 Control Plane, 1 etcd, 3 workers

**Describe the bug:**

When provisioning clusters with local cluster authentication endpoint enabled and multiple nodes utilizing explicit control plane roles only one of the secondary contexts is reachable, the other(in a 2 control plane cluster) responed with "Unauthorized"

**Steps To Reproduce:**

1) Provision a node driver Amazon EC2 RKE2 cluster with multiple node pools dedicated to the control plane role
	Pool 0 - 1 Control Plane
	Pool 1 - 1 Control Plane
	Pool 2 - 1 Etcd
	Pool 3 - 1 Worker
	Pool 4 - 1 Worker
	Pool 5 - 1 Worker

2) Enabled Local Cluster Authenticaiton Endpoint, leave FQDN and CACERT blank
	View cluster's YAML and verify localClusterAuthEndpoint = true

3) Download KubeConfig for provisoned cluster view the available contexts (one for each control plane node)
	kubectl config get-contexts

3) Access cluster on each listed context (Unauthorized response)
	kubectl config use-context xxxx-xxxx
	kubectl get pods -A

**Expected behavior:**

An accessible secondary context for each control plane by enabling local cluster auth endpoint

**Actual behavior:**

Consistently unable to reach secondary context in clusters with multiple explicit control plane roles

**Additional context / logs:**

| RKE2 Verison | Single Secondary Context | Multiple Contexts | Notes |
| ------------ | -------------------------| ----------------- | ----- |
| 1.21.9 | Failed | Failed | Showing Enabled in UI |
| 1.21.8+rke2r2 | Passed | Failed | One secondary context available |
| 1.22.6 | Pass | Pass/Failed | Consistently seeing only one secondary context availble, intermittent |
| 1.23.3 | Pass | Pass/Failed | Consistently seeing only one secondary context availble, intermittent |

**Comment by thedadams** [comment]
There was a change made that Only started clusterauthtoken controller if ACE is enabled. This is where I would start investigating this issue.

https://github.com/rancher/rancher/pull/35520

**Comment by paynejacob** [comment]
For who investigates this, check if the non working context is using the etcd node.

**Comment by jiaqiluo** [comment]
A DaemonSet `kube-api-auth` will be deployed to all control plane nodes if ACE is enabled.  
The request will go through the `kube-api-auth` pod on the corresponding node when we use the secondary context in the kubectl command. 

In the logs of the `kube-api-auth` pod, we see the following kind of error:
```
time="2022-03-03T21:43:42Z" level=info msg="  ...looking up token for kubeconfig-user-j8pg6zfhd4"
time="2022-03-03T21:43:42Z" level=error msg="clusteruserattributes.cluster.cattle.io \"cattle-system/user-j8pg6\" not found"
``` 
 
Also, the logs prints out all controllers the pod starts, but the following message is missing in the logs of the broken pod which indicates the controller for `ClusterUserAttribute`  is not started:
```
time="2022-03-02T20:49:22Z" level=info msg="Starting cluster.cattle.io/v3, Kind=ClusterUserAttribute controller" 
```

Restarting the pod can mitigate the problem but not all the time. Sometimes we still see the error after restarting the pod. 

We thought this bug might be related to[an issue in lasso](https://github.com/rancher/lasso/pull/13) related to staled cache, so I attempted to bump the versions of lasso and rancher used in the `kube-api-auth`.  However, it did not resolve the problem fully: the problem still happens on some control plane nodes sometimes. 

A further investigation is needed.

**Comment by snasovich** [comment]
Moving to 2.6.5 milestone given the intermittent nature of the issue and the fact it doesn't seem to be a regression in 2.6-head compared to 2.6.3 (so this issue was there for a while).

**Comment by jiaqiluo** [comment]
### Root cause
The bug happens in the `kube-api-auth` server which runs in the `kube-api-auth` pod.  
Sometimes the ClusterUserAttribute controller is not started, which causes the failure to retrieve the ClusterUserAttribute for the current user and subsequently the failure to validate the cluster auth token. Lasso has the behavior that it will delay the start of a controller if there is no listener or handler associated with it, and in our case, it might cause the controller never to start.


### What was fixed, or what changes have occurred
So the fix is to move the registration of listeners before starting the controller, which makes sure the controller is started immediately.
In rancher, we bump the image tag of kube-api-auth to v0.1.8 which contains the fix.
 
### Areas or cases that should be tested
 We need to test if the secondary control plane context works for the RKE/RKE2 cluster which has multiple control plane nodes and the ACE is enabled, and they should work when the `rancher/kube-api-auth:v0.1.8` image is used. 
 
### What areas could experience regressions?
n/a  

### Are the repro steps accurate/minimal?
yes

**Comment by jiaqiluo** [comment]
the fix is available in rancher:v2.6-head 0ad0def

**Comment by markusewalker** [comment]
Verified in v2.6-head https://github.com/rancher/rancher/commit/c8114a946563c4ca902881de703c6e4b55b72cb5.

**Environments Tested**
- Amazon EC2 node driver cluster (provisioned in Rancher UI)
     - RKE version: `v1.22.8-rancher1-1`
          - 6 Node Pools: 1 etcd, 3 control planes, 3 workers
     - RKE2 version: `v1.22.8+rke2r1`
          - 6 Node Pools: 1 etcd, 2 control planes, 3 workers
- Linode node driver cluster (provisioned in Rancher UI)
     - RKE version: `v1.22.8-rancher1-1`
          - 6 Node Pools: 1 etcd, 3 control planes, 3 workers
     - RKE2 version: `v1.22.8+rke2r1`
          - 6 Node Pools: 1 etcd, 2 control planes, 3 workers

**Results**

NODE DRIVER | RKE (PASS/FAIL) | RKE2 ( PASS/FAIL)
-- | -- | --
Amazon EC2 | PASS | PASS
Linode (1 etcd, 2   control plane, 3 workers) | PASS | PASS

**Verification Steps**
1. Provisioned downstream Linode cluster in Rancher UI for RKE.
2. Downloaded kubeconfig file for cluster and appended to local client machine's `$HOME/.kube` location.
3. Ran the following commands:

```
$ kubectl config get-contexts
CURRENT   NAME                          CLUSTER                       AUTHINFO                NAMESPACE
*         linode-36518-rke-test         linode-36518-rke-test         linode-36518-rke-test   
          linode-36518-rke-test-node3   linode-36518-rke-test-node3   linode-36518-rke-test   
          linode-36518-rke-test-node5   linode-36518-rke-test-node5   linode-36518-rke-test   
          linode-36518-rke-test-node6   linode-36518-rke-test-node6   linode-36518-rke-test   
$ kubectl config use-context linode-36518-rke-test-node3
Switched to context "linode-36518-rke-test-node3".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             15m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (15m ago)   17m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
$ kubectl config use-context linode-36518-rke-test-node5
Switched to context "linode-36518-rke-test-node5".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             15m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (15m ago)   18m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
$ kubectl config use-context linode-36518-rke-test-node6
Switched to context "linode-36518-rke-test-node6".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             16m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (16m ago)   18m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
```
4. Repeated steps 1-3, but this time used RKE2. Similar commands gave the below results:

```
$ kubectl get nodes
NAME                                   STATUS   ROLES                  AGE     VERSION
36518-rke2-test-pool1-0866e6aa-ljll6   Ready    control-plane,master   11m     v1.22.8+rke2r1
36518-rke2-test-pool2-f22f0ed1-rdnr4   Ready    control-plane,master   10m     v1.22.8+rke2r1
36518-rke2-test-pool3-eef9e0d2-ghs7q   Ready    etcd                   11m     v1.22.8+rke2r1
36518-rke2-test-pool4-0fe7cb8b-wlgd6   Ready    worker                 6m19s   v1.22.8+rke2r1
36518-rke2-test-pool5-8ac34db8-d4m5z   Ready    worker                 6m27s   v1.22.8+rke2r1
36518-rke2-test-pool6-94fefb7f-gnp2q   Ready    worker                 6m5s    v1.22.8+rke2r1
$ kubectl config get-contexts
CURRENT   NAME                                   CLUSTER                                AUTHINFO          NAMESPACE
*         36518-rke2-test                        36518-rke2-test                        36518-rke2-test   
          36518-rke2-test-pool1-0866e6aa-ljll6   36518-rke2-test-pool1-0866e6aa-ljll6   36518-rke2-test   
          36518-rke2-test-pool2-f22f0ed1-rdnr4   36518-rke2-test-pool2-f22f0ed1-rdnr4   36518-rke2-test   
$ kubectl config use-context 36518-rke2-test-pool1-0866e6aa-ljll6
Switched to context "36518-rke2-test-pool1-0866e6aa-ljll6".
$ kubectl get pods -A
NAMESPACE             NAME                                                            READY   STATUS      RESTARTS   AGE
calico-system         calico-kube-controllers-6487c68d68-45f5r                        1/1     Running     0          10m
calico-system         calico-node-7wrfd                                               1/1     Running     0          7m4s
calico-system         calico-node-bv48s                                               1/1     Running     0          10m
calico-system         calico-node-jc6xp                                               1/1     Running     0          6m56s
...

$ kubectl config use-context 36518-rke2-test-pool2-f22f0ed1-rdnr4
Switched to context "36518-rke2-test-pool2-f22f0ed1-rdnr4".
$ kubectl get pods -A
NAMESPACE             NAME                                                            READY   STATUS      RESTARTS   AGE
calico-system         calico-kube-controllers-6487c68d68-45f5r                        1/1     Running     0          11m
calico-system         calico-node-7wrfd                                               1/1     Running     0          7m44s
calico-system         calico-node-bv48s                                               1/1     Running     0          11m
calico-system         calico-node-jc6xp                                               1/1     Running     0          7m36s
...
```
5. Repeated steps 1-3, but for Amazon EC2 for RKE1. Ran the below commands:

```
$ kubectl config get-contexts
CURRENT   NAME                          CLUSTER                       AUTHINFO                NAMESPACE
*         linode-36518-rke-test         linode-36518-rke-test         linode-36518-rke-test   
          linode-36518-rke-test-node3   linode-36518-rke-test-node3   linode-36518-rke-test   
          linode-36518-rke-test-node5   linode-36518-rke-test-node5   linode-36518-rke-test   
          linode-36518-rke-test-node6   linode-36518-rke-test-node6   linode-36518-rke-test   
$ kubectl config use-context linode-36518-rke-test-node3
Switched to context "linode-36518-rke-test-node3".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             15m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (15m ago)   17m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
$ kubectl config use-context linode-36518-rke-test-node5
Switched to context "linode-36518-rke-test-node5".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             15m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (15m ago)   18m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
$ kubectl config use-context linode-36518-rke-test-node6
Switched to context "linode-36518-rke-test-node6".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS      AGE
cattle-cluster-agent-7874b78576-pwg7t   1/1     Running   0             16m
cattle-cluster-agent-7874b78576-s9dr2   1/1     Running   1 (16m ago)   18m
cattle-node-agent-2xrmb                 1/1     Running   0             18m
...
```
6. Repeated steps 1-3. but for Amazon EC2 for RKE2. Ran the below commands:

```
$ kubectl get nodes
NAME                                               STATUS   ROLES                  AGE   VERSION
markus-36518-testing-markus-pool1-97d03ab6-7dbv6   Ready    etcd                   19m   v1.22.8+rke2r1
markus-36518-testing-markus-pool2-322f8eb2-4hjwv   Ready    control-plane,master   19m   v1.22.8+rke2r1
markus-36518-testing-markus-pool3-c77e27a2-77p29   Ready    control-plane,master   19m   v1.22.8+rke2r1
markus-36518-testing-markus-pool4-8eb5e5d2-4wkx4   Ready    worker                 15m   v1.22.8+rke2r1
markus-36518-testing-markus-pool5-d249cbe0-hxjnq   Ready    worker                 15m   v1.22.8+rke2r1
markus-36518-testing-markus-pool6-e7217685-mmbs9   Ready    worker                 15m   v1.22.8+rke2r1
$ kubectl config get-contexts
CURRENT   NAME                                               CLUSTER                                            AUTHINFO               NAMESPACE
*         markus-36518-testing                               markus-36518-testing                               markus-36518-testing   
          markus-36518-testing-markus-pool2-322f8eb2-4hjwv   markus-36518-testing-markus-pool2-322f8eb2-4hjwv   markus-36518-testing   
          markus-36518-testing-markus-pool3-c77e27a2-77p29   markus-36518-testing-markus-pool3-c77e27a2-77p29   markus-36518-testing   
$ kubectl config use-context markus-36518-testing-markus-pool2-322f8eb2-4hjwv
Switched to context "markus-36518-testing-markus-pool2-322f8eb2-4hjwv".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS   AGE
cattle-cluster-agent-689bf4d789-rxlxl   1/1     Running   0          17m
cattle-cluster-agent-689bf4d789-w9v94   1/1     Running   0          17m
kube-api-auth-kbkbn                     1/1     Running   0          18m
kube-api-auth-ltdcg                     1/1     Running   0          18m
$ kubectl config use-context markus-36518-testing-markus-pool3-c77e27a2-77p29
Switched to context "markus-36518-testing-markus-pool3-c77e27a2-77p29".
$ kubectl get pods -n cattle-system
NAME                                    READY   STATUS    RESTARTS   AGE
cattle-cluster-agent-689bf4d789-rxlxl   1/1     Running   0          17m
cattle-cluster-agent-689bf4d789-w9v94   1/1     Running   0          17m
kube-api-auth-kbkbn                     1/1     Running   0          19m
kube-api-auth-ltdcg                     1/1     Running   0          19m
```

I did note that in specifically RKE, when looking to verify if `localClusterAuthEndpoint` is set to true, I get a blank value:

![image](https://user-images.githubusercontent.com/52591819/164087263-4ea7a77b-c438-449a-83c9-c78b32630074.png)

RKE2 did not have this issue:

![image](https://user-images.githubusercontent.com/52591819/164087330-a5f17794-428a-48aa-95b4-4369076dda6e.png)

@jiaqiluo is this expected behavior for it to be empty for RKE? If not, we can make a separate issue for that and close this one accordingly. Please let me know.

**Comment by thedadams** [comment]
@markusewalker Yes, the localClusterAuth should be true on the management cluster object in both cases.
